File: Inheritance/2.rb
#puts 대신 p를 사용합니다. p는 함수이고, ()을 생략해준다.
# subtract = 빼는 값
#//생성자, constructor
#initialize = 초기화 약속된 메소드
#instance 값을 끌어 들어왔다.
# (10,10) @뒤에 오는것을 인스턴스 v1이다.
#변수이름앞에 @를 붙이면 됩니다.
#똑같은 객체에 속해 있기 때문에 연관성이 잘들어나있다.
#루비는 인스턴스값에 직접적으로 접근할수 없다. 그러기때문에 getv1 메소드를 사용해야한다.
class Cal
  def initialize(v1, v2)
    p v1, v2
    @v1 = v1
    @v2 = v2
  end
  def add()
    return @v1+@v2
  end
  def subtract()
    return @v1-@V2
  end
  def setv1(v)
    if v.is_a?(Integer)
      @v1 = v
    end
  end
  def getv1()
    return @v1
  end
end
class CalMultiply < Cal
  def multiply()
    return @v1*@v2
  end
end
class CalDivide < CalMultiply
    return @v1/@v2
  end
end
c1 = CalMultiply.new(10,10)
p c1.add()
p c1.multiply()
c2 = CalMultiply.new(20, 10)
p c2, c2.add()
p c2, c2.multiply()
p c2, c2.divide()

# c1 = Cal.new(10,10) #c1 인스턴드안에 매소드와 인스턴스변수가 아주 안전하게 보관되어있다. 다른코드와 변할수있는 확률이 적다
# p c1.add()
# p c1.subtract()
# c1.setv1('one')
# p c1.add()

File: Multiple_Inheritance/2.py
#method = function
#파이썬의 메소드는 첫번째 매개변수를 정의해야합니다.
#첫번째 매게변수는 언제나 그 인스턴스가 됩니다. 그 인스턴스를 값을 매개변수를 가리킵니다.
#메소드를 호출할때 첫번째 매개변수로 첫번째 매개변수로 들어오는것을 약속되어있기때문에 첫번째 매개변수를 이용해서 인스턴스를 접근할수있다.
#self = s 변수의 이름
#cal을 CalMultiply,CalDivide로 사용했다.

class CalMultiply():
    def multiply(self):
        return self.v1*self.v2
class CalDivide():
    def divide(self):
        return self.v1/self.v2
class Cal(CalMultiply, CalDivide):
    def __init__(self, v1, v2):
        if isinstance(v1, int):
            self.v1 = v1
        if isinstance(v2, int):
            self.v2 = v2
    def add(self):
        return self.v1+self.v2
    def subtract(self):
        return self.v1-self.v2
    # def multiply(self):
    # return self.v1*self.v2
    def setv1(self, v):
        if isinstance(v, int):                 #만약 v가 숫자라면:
        self.v1 = v
    def getv1(self):
        return self.v1
    c= Cal(100, 10)
    print(c, add())
    print(c. multiply())
    print(c. divide())


c1 = CalMultiply(10,10)
print(c1.add())
print(c1.multiply())
c2 = CalDivide(20, 10)
print(c2, c2.add())
print(c2, c2.multiply())
print(c2, c2.divide())


# c1 = Cal(10,10)
# print(c1.add())
# print(c1.subtract())
# c1.setv1 = ('one')
# c1.v2 = 30
# print(c1.add())
# print(c1.subtract())
